Source/FieldSolver/WarpXPushFieldsEM.cpp

```diff
@@ -243,11 +262,41 @@ WarpX::PSATDForwardTransformJ ()
 
     for (int lev = 0; lev <= finest_level; ++lev)
     {
-        ForwardTransformVect(lev, *spectral_solver_fp[lev], current_fp[lev], idx_jx, idx_jy, idx_jz);
+        // With Vay's deposition, J stores a temporary current (D) that is later modified
+        // in Fourier space: its staggering is always nodal and does not match that of J
+        amrex::IntVect jx_stag =
+            (WarpX::current_deposition_algo == CurrentDepositionAlgo::Vay) ?
+            amrex::IntVect::TheNodeVector() : current_fp[lev][0]->ixType().toIntVect();
+
+        amrex::IntVect jy_stag =
+            (WarpX::current_deposition_algo == CurrentDepositionAlgo::Vay) ?
+            amrex::IntVect::TheNodeVector() : current_fp[lev][1]->ixType().toIntVect();
+
+        amrex::IntVect jz_stag =
+            (WarpX::current_deposition_algo == CurrentDepositionAlgo::Vay) ?
+            amrex::IntVect::TheNodeVector() : current_fp[lev][2]->ixType().toIntVect();
+
+        ForwardTransformVect(lev, *spectral_solver_fp[lev], current_fp[lev],
+                             idx_jx, idx_jy, idx_jz, jx_stag, jy_stag, jz_stag);
 
         if (spectral_solver_cp[lev])
         {
-            ForwardTransformVect(lev, *spectral_solver_cp[lev], current_cp[lev], idx_jx, idx_jy, idx_jz);
+            // With Vay's deposition, J stores a temporary current (D) that is later modified
+            // in Fourier space: its staggering is always nodal and does not match that of J
```

Comment:
More specifically, it seems to me that the staggering of `D` should match that of `rho` (which is currently always fully nodal, but could at some point be fully cell-centered). Could you modify the current code to define `jx_stag` based on the index type of `rho_fp`/`rho_cp`?